### PR TITLE
Back Office web frontend Azure AD authentication

### PR DESCRIPTION
**Pull request by lestermarch**
- Add Azure AD authentication support to the Node App Service module.
- Add auth-related App Service settings for Back Office web frontend.

### CODE REVIEW

File: config/variables/stacks/back-office/dev.hcl [hunk 1]
```diff
@@ -1,4 +1,10 @@
 locals {
+  azuread_auth_case_officer_group_id      = "c921094a-318f-4996-be5e-9bd2ef9b7bdf"
+  azuread_auth_client_id                  = "2dc22997-3900-4602-b3bd-46385b60e2b2"
+  azuread_auth_enabled                    = true
```
Comment (andylangridge) [inline config/variables/stacks/back-office/dev.hcl:4]: As per Radu's comments I don't think this needs to be set as a variable for the component/stack. Can simply just pass in `true` to the module :)

Comment (lestermarch) [reply config/variables/stacks/back-office/dev.hcl]: Fixed in https://github.com/Planning-Inspectorate/infrastructure-environments/pull/80/commits/cb7ea0166ba6669b373478723805838979484c04

File: app/modules/node-app-service/main.tf [hunk 2]
```diff
@@ -97,6 +97,43 @@ resource "azurerm_linux_web_app_slot" "staging" {
   }
 }
 
+resource "azurerm_app_service_certificate" "custom_hostname" {
+  count = var.custom_hostname_enabled ? 1 : 0
```
Comment (andylangridge) [inline app/modules/node-app-service/main.tf:101]: Maybe we don't need this flag variable? Can we do a `var.custom_hostname != null` check instead?

File: app/modules/node-app-service/main.tf [hunk 3]
```diff
@@ -97,6 +97,43 @@ resource "azurerm_linux_web_app_slot" "staging" {
...
+  tags = var.tags
+}
+
+resource "azurerm_app_service_custom_hostname_binding" "custom_hostname" {
+  count = var.custom_hostname_enabled ? 1 : 0
+
+  hostname            = var.custom_hostname
+  app_service_name    = azurerm_linux_web_app.web_app.name
+  resource_group_name = var.resource_group_name
+  # thumbprint          = azurerm_app_service_certificate.custom_hostname[0].thumbprint
```
Comment (andylangridge) [inline app/modules/node-app-service/main.tf:117]: Can we remove this if we're not using? :)

**Review by andylangridge — APPROVED** [review]

File: app/modules/node-app-service/main.tf (hunk 2 above)
Comment (lestermarch) [reply app/modules/node-app-service/main.tf]: Fixed in [ec6468c](https://github.com/Planning-Inspectorate/infrastructure-environments/pull/80/commits/ec6468ccebb29a58feeec51e6ee65de9e51c126a)

File: app/stacks/global/front-door/front-door.tf [hunk 4]
```diff
@@ -92,7 +92,7 @@ resource "azurerm_frontdoor" "common" {
       backend {
         enabled     = true
         address     = mapping.value["app_service_url"]
-        host_header = mapping.value["app_service_url"]
+        host_header = mapping.value["infer_backend_host_header"] == false ? mapping.value["app_service_url"] : ""
```
[the remark this answers is not in the document]
Comment (lestermarch) [reply app/stacks/global/front-door/front-door.tf]: Fixed in [ec6468c](https://github.com/Planning-Inspectorate/infrastructure-environments/pull/80/commits/ec6468ccebb29a58feeec51e6ee65de9e51c126a)

File: app/modules/node-app-service/main.tf (hunk 3 above)
Comment (lestermarch) [reply app/modules/node-app-service/main.tf]: Comments removed in subsequent commits.

**Review by andylangridge — APPROVED** [review]

**Pull request merged**